src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+            var bestItem = Session.GetBestCompletionItemBasedOnMRU(chosenItems, recentItems);
+            VSCompletionItem uniqueItem = null;
+            int selectedItemIndex = 0;
+
+            // Determine if we should consider this item 'unique' or not.  A unique item
+            // will be automatically committed if the user hits the 'invoke completion' 
+            // without bringing up the completion list.  An item is unique if it was the
+            // only item to match the text typed so far, and there was at least some text
+            // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+            // like "WriteLine" since no filter text has actually been provided.  HOwever,
+            // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+            if (bestItem != null)
+            {
+                selectedItemIndex = itemsInList.IndexOf(i => Equals(i.FilterResult.CompletionItem, bestItem));
+                if (selectedItemIndex > -1 && bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
+
+            // If we don't have a best completion item yet, then pick the first item from the list.
+            var bestOrFirstCompletionItem = bestItem ?? itemsInList.First().FilterResult.CompletionItem;
+
+            // Check that it is a filter symbol. We can be called for a non-filter symbol.
+            if (filterReason == CompletionFilterReason.Insertion &&
+                !Controller.IsPotentialFilterCharacter(typeChar) &&
+                !string.IsNullOrEmpty(filterText) &&
+                !Controller.IsFilterCharacter(bestOrFirstCompletionItem, typeChar, filterText))
+            {
+                return null;
+            }
+
+            var updateSelectionHint = Session.IsHardSelection(
+                        filterText, initialRoslynTriggerKind, bestOrFirstCompletionItem,
+                        _completionHelper, filterReason, recentItems, hasSuggestedItemOptions)
+                        ? AsyncCompletionData.UpdateSelectionHint.Selected
+                        : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                highlightedList, selectedItemIndex, filters,
+                updateSelectionHint, centerSelection: true, uniqueItem);
+        }
+
+        private AsyncCompletionData.FilteredCompletionModel HandleDeletionTrigger(
+            List<ExtendedFilterResult> filterResults,
+            string filterText,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters,
+            ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight> highlightedList)
+        {
```

Comment:
Where is the logic for:

```
                    // The user has typed something, but nothing in the actual list matched what
                    // they were typing.  In this case, we want to dismiss completion entirely.
                    // The thought process is as follows: we aggressively brough up completion
                    // to help them when they typed delete (in case they wanted to pick another
                    // item).  However, they're typing something that doesn't seem to match at all
                    // The completion list is just distracting at this point.
```
?

Reply:
I think the most of the scenario should be covered by HandleAllItemsFilteredOut. The only difference I see is where filterText.Length <= 1.

Added [https://github.com/dotnet/roslyn/issues/31135](https://github.com/dotnet/roslyn/issues/31135) to check if we need to specifically cover the case.

---
In reply to: [232510424](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510424) [](ancestors = 232510424)